src/Symfony/Component/HttpFoundation/InputBag.php

```diff
@@ -0,0 +1,122 @@
+    {
+        if (null !== $default && !is_scalar($default) && !\is_callable([$default, '__toString'])) {
+            @trigger_error(sprintf('Passing a non-string value as 2nd argument to "%s()" is deprecated since Symfony 5.1, pass a string or null instead.', __METHOD__), E_USER_DEPRECATED);
+        }
+
+        $value = parent::get($key, $this);
+
+        if (null !== $value && $this !== $value && !is_scalar($value) && !\is_callable([$value, '__toString'])) {
+            @trigger_error(sprintf('Retrieving a non-string value from "%s()" is deprecated since Symfony 5.1, and will throw a "%s" exception in Symfony 6.0, use "%s::all()" instead.', __METHOD__, BadRequestException::class, __CLASS__), E_USER_DEPRECATED);
+        }
+
+        return $this === $value ? $default : $value;
+    }
+
+    /**
+     * Returns the inputs.
+     *
+     * @param string|null $key The name of the input to return or null to get them all
+     */
+    public function all(string $key = null): array
+    {
+        if (null === $key) {
+            return $this->parameters;
+        }
+
+        $value = $this->parameters[$key] ?? [];
+        if (!\is_array($value)) {
+            throw new Exception\BadRequestException(sprintf('Unexpected value for "%s" input, expecting "array", got "%s".', $key, \is_object($value) ? \get_class($value) : \gettype($value)));
+        }
+
+        return $value;
+    }
+
+    /**
+     * Replaces the current input values by a new set.
+     */
+    public function replace(array $inputs = [])
+    {
+        $this->parameters = [];
+        $this->add($inputs);
+    }
+
+    /**
+     * Adds input values.
+     */
+    public function add(array $inputs = [])
+    {
+        foreach ($inputs as $input => $value) {
+            $this->set($input, $value);
+        }
+    }
+
+    /**
+     * Sets an input by name.
+     *
+     * @param string|array $value
+     */
+    public function set(string $key, $value)
+    {
+        if (!is_scalar($value) && !\is_callable([$value, '__toString']) && !\is_array($value)) {
+            @trigger_error(sprintf('Passing "%s" as a 2nd Argument to "%s()" is deprecated since Symfony 5.1, pass a string or an array instead.', \is_object($value) ? \get_class($value) : \gettype($value), __METHOD__), E_USER_DEPRECATED);
+        }
+
+        $this->parameters[$key] = $value;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function filter(string $key, $default = null, int $filter = FILTER_DEFAULT, $options = [])
+    {
+        $value = $this->has($key) ? $this->all()[$key] : $default;
+
+        // Always turn $options into an array - this allows filter_var option shortcuts.
+        if (!\is_array($options) && $options) {
+            $options = ['flags' => $options];
+        }
+
+        // Add a convenience check for arrays.
+        if (\is_array($value)) {
+            if (!(($options['flags'] ?? 0) & (FILTER_REQUIRE_ARRAY | FILTER_FORCE_ARRAY))) {
```

Comment:
to be merged with the previous "if"

Reply:
there's another `if` statement that probably should not be removed. see https://github.com/symfony/symfony/pull/34363#discussion_r376446871